cpu/esp_common/esp-now/esp_now_gnrc.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (C) 2018 Timo Rothenpieler
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_esp_common_esp_now
+ * @{
+ *
+ * @file
+ * @brief       Netif interface for the ESP-NOW WiFi P2P protocol
+ *
+ * @author Timo Rothenpieler <timo.rothenpieler@uni-bremen.de>
+ */
+
+#include <assert.h>
+#include <stdlib.h>
+
+#include <sys/uio.h>
+
+#include "net/netdev.h"
+#include "net/gnrc.h"
+#include "esp_now_params.h"
+#include "esp_now_netdev.h"
+#include "esp_now_gnrc.h"
+#include "net/gnrc/netif.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static int _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt)
+{
+    uint8_t mac[ESP_NOW_ADDR_LEN];
+    uint8_t plbuf[ESP_NOW_MAX_SIZE_RAW];
+    netdev_t *dev = netif->dev;
+
+    assert(pkt != NULL);
+
+    gnrc_netif_hdr_t *netif_hdr;
+    gnrc_pktsnip_t *payload;
+
+    payload = pkt->next;
+
+    if (pkt->type != GNRC_NETTYPE_NETIF) {
+        DEBUG("gnrc_esp_now: First header was not generic netif header\n");
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    netif_hdr = (gnrc_netif_hdr_t*)pkt->data;
+
+    if (netif_hdr->flags & (GNRC_NETIF_HDR_FLAGS_BROADCAST | GNRC_NETIF_HDR_FLAGS_MULTICAST)) {
+        /* ESP-NOW does not support multicast, always broadcast */
+        memset(mac, 0xff, ESP_NOW_ADDR_LEN);
+    } else if (netif_hdr->dst_l2addr_len == ESP_NOW_ADDR_LEN) {
+        memcpy(mac, gnrc_netif_hdr_get_dst_addr(netif_hdr), ESP_NOW_ADDR_LEN);
+    } else {
+        DEBUG("gnrc_esp_now: destination address had unexpected format"
+              "(flags=%d, dst_l2addr_len=%d)\n", netif_hdr->flags, netif_hdr->dst_l2addr_len);
+        gnrc_pktbuf_release(pkt);
+        return -EBADMSG;
+    }
+
+    iolist_t payload_iolist = {
+        .iol_base = plbuf,
+        .iol_len = sizeof(esp_now_pkt_hdr_t)
+    };
+
+    iolist_t iolist = {
+        .iol_base = mac,
+        .iol_len = sizeof(mac),
+        .iol_next = &payload_iolist
```

Comment:
Since `iolist_t` has (intentionally) the same structure as `gnrc_pktsnip_t` (i.e. is a super class) you can just do
```suggestion
         .iol_next = (iotlist_t *)payload
```
and safe yourself a lot of copying work above and below.

Reply:
(see e.g. `gnrc_netif_ethernet` implementation).

Reply:
That's unfortunately not possible, since the esp_now_pkt_hdr has to be inserted in front of the data, and the underlying esp_now functions for sending need to have all that data in one continuous buffer.

Reply:
That's unfortunate. Then let's fix that later. This is a huge memory (and time) waste that shouldn't be necessary. Maybe it is the reason for the easily overflowing buffer (every packet needs more than twice its size in packet buffer space).

Reply:
I put quite some thought into how to optimize this specific part, and ended up at the conclusion that at least one full copy of the data to be sent is unavoidable with how the esp_now API works.
The memory its copied into is on the stack of the netifs _send function, so I doubt this is related to the overfilling buffers.

Reply:
Maybe not the overfilling buffer, but a stack overflow ;-).

Reply:
@TimoRoth can you maybe reconstruct your train of thought for this, because I don't see, how copying the whole buffer is more optimized, than just reusing a pointer.

Reply:
To explain a bit more what's happening in that code:
The esp_now netdevs send function expects a very specific iolist setup to be passed to it. The first element of the list contains the destination mac address, and the second element the complete buffer to be sent, which is then passed to esp_now_send() untouched.
If I were to pass a multi-element iolist there instead, it itself would have to copy around the data, ending up with the same overhead.

Reply:
It's not just copying the whole buffer, it is also prepending a header and combining the potentially multiple payload iolist elements into one single buffer, which is required by the API.

Reply:
But in the esp_now's send function you would be able to use the state of the device for buffer space (i.e. non-stack space, not risking stack overflows) instead of the stack. With the restricted stack sizes we typically have, putting buffers into the stack is always a alarm bell that should ring in the back of your head.

Reply:
Also you wouldn't put a constraint on the `iolist` argument of the send function which other devices don't do.

Reply:
Done pretty much exactly that now.